crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -1,83 +1,764 @@
+
+        for (path, version) in packages {
+            let mut value = rnix::ast::AttrSet::new();
+            if let Some(version) = version {
+                value = value.insert_unchecked(
+                    ["version"],
+                    rnix::ast::Str::new(&version).syntax().to_owned(),
+                );
+            }
+
+            let mut path_in_packages = vec!["packages".to_string()];
+            path_in_packages.extend_from_slice(&path);
+            edited = edited.insert_unchecked(path_in_packages, value.syntax().to_owned());
+        }
+
+        let green_tree = config_attrset
+            .syntax()
+            .replace_with(edited.syntax().green().into_owned());
+        let new_content = nixpkgs_fmt::reformat_string(&green_tree.to_string());
+
+        self.set_environment(new_content.as_bytes(), nix, system)
+            .await
+    }
+
+    /// Uninstall packages by converting a [FloxPackage]s into attributes in the `flox.nix` format,
+    /// and then using [`rnix`](https://crates.io/crates/rnix) to remove these attributes from the
+    /// environment definition file.
+    pub async fn uninstall(
+        &mut self,
+        packages: impl IntoIterator<Item = FloxPackage>,
+        nix: &NixCommandLine,
+        system: impl AsRef<str>,
+    ) -> Result<&mut Self, EnvironmentError2> {
+        let packages = packages
+            .into_iter()
+            .map(|package| package.flox_nix_attribute().unwrap());
+
+        // assume flake exists locally
+        let flox_nix_path = self.flox_nix_path();
+        let flox_nix_content: String = std::fs::read_to_string(&flox_nix_path).unwrap();
+
+        let mut root = rnix::Root::parse(&flox_nix_content)
+            .ok()
+            .unwrap()
+            .expr()
+            .unwrap();
+
+        if let Expr::Lambda(lambda) = root {
+            root = lambda.body().unwrap();
+        }
+
+        let config_attrset = find_attrs(root.clone()).unwrap();
+        let mut edited = config_attrset.clone().syntax().green().into_owned();
+
+        for (path, _version) in packages {
+            let mut path_in_packages = vec!["packages".to_string()];
+            path_in_packages.extend_from_slice(&path);
+
+            let index = config_attrset
+                .find_by_path(&path_in_packages)
+                .unwrap_or_else(|| panic!("path not found, {path_in_packages:?}"))
+                .syntax()
+                .index();
+            edited = edited.remove_child(index - 2); // yikes
+        }
+
+        let green_tree = config_attrset.syntax().replace_with(edited);
+        let new_content = nixpkgs_fmt::reformat_string(&green_tree.to_string());
+
+        self.set_environment(dbg!(new_content).as_bytes(), nix, system)
+            .await
+    }
+
+    /// Commmit changes, by moving modified files back to the original (read only) location
+    pub fn finish(self) -> Result<Environment<Read>, EnvironmentError2> {
```

Comment:
I'm not sure this should be the responsibility of the `Environment`. I would expect this to be done by the `flox edit` command. I see the `Environment` as mostly a container, and it feels like a boundary violation to allow the container to overwrite the origin.

Reply:
Would you want to move this method to `Environment<Read>`? If we do so it would probably make more sense to name them `Environment<Original>` and `Environment<Temporary>`?

Reply:
I also think in some cases it would make more sense to have methods like install on `Environment<Original>` even though they modify the environment. E.g. it would be easier for flox CLI to consume `install` if it didn't have to worry about calling `modify_in`. In `Environment<Original>::install` we could have:
```
temporary = modify_in()
temporary.set_environment() // we could even have set_environment container the modify_in call
self.finish(temporary)
```

That simplifies the API (just call install, rather than calling modify_in and install), and I don't think we lose anything in terms of transactionality

Reply:
I think it makes less sense for `install` to be on `Environment<Read>` because I read the `State` as what permissions you have, and it doesn't make sense to me to be able to install something when I have read-only permissions.

I like the `Original`/`Temporary` distinction. In that case I would see a flow going like this:
```rust
fn try_install(self: Environment<Original>, pkg: Package) -> Result<(), SomeError> {
    let mut temp_env: Environment<Temporary> = self.make_temporary()?;
    let new_contents = temp_env.contents_with_new_package(pkg)?;
    temp_env.update_file(new_contents);
    let succeeded = temp_env.build()?;
    if succeeded {
        self.replace_with(temp_env)?;
        return Ok(());
    } else {
        temp_env.clean_up()?;
        bail!("could not install package")
    }
}
```

Reply:
Agree a split `Original`/`Temporary` and only consume methods from `Original` seem useful.

Some concerns:

* should `flox install a b c` (b being invalid) a) fail atomically b) install `a` c) install `a` & `c`? 
  `try_install` taking a single package we can't provide a). In remote envs (considering this might be a trait) we'd also generate multiple generations (one per call to try_install)
  I dont think we have to block on any of that now, seems simple enough to change.
  
* We still need to pass in the tempdir at some point, either store it in `Original` or pass as argument

* You prefer methods like `contents_with_new_package` that would just give you an updated str rather than editing the (temporary) rather than `contents_with_new_package` calling `update_file` itself?

Reply:
To separate things a bit more I previously thought of using A pattern closer to `Publish`

* a trait for environments exposing _readable_ fields (path, installable, owner, name, packages, etc, build(?) )

* a struct for the impl of modify envs :

``` rust
struct ModifyEnv<E> {
	original: E,
    temporary: LocalEnv
}

impl <E: Environment> ModifyEnv<E> {
    fn env_declaration_with_new_package(&self) -> Result<String>; 
    fn env_declaration_without_package(&self) -> Result<String>;

	fn set_env_declaration_content(&mut self, String) -> Result<()>; 
    fn try_build(&self) -> Result<()>;
}

// apply changes, differently per "Environment backend"

impl ModifyEnv<LocalEnv> {
    fn apply(self) -> Result<LocalEnv>;
}


impl ModdifyEnv<RemoteEnv> {
    fn apply(self) -> Result<RemoteEnv>;
}
```

* A struct to impl activate 
* A struct for push/pull operations


this way we should be able to work on separate subsystems without having to touch the same struct or trait that much.
This would also make it easier to implement some parts in other languages through json interfaces without pulling all that into the environment itself.